Update get campaigns to return targeted locations

ianlin opened the pull request:
### Changes proposed in this Pull Request:

Part of #1248, this PR updates two endpoints `GET ads/campaigns` and `GET ads/campaigns/:campaign-id` to have `targeted_locations` in the response payload in order to support multi-country targeting. Multiple targeted locations can be retrieve from [CampaignCriterion](https://developers.google.com/google-ads/api/reference/rpc/v9/CampaignCriterion)'s [LocationInfo](https://developers.google.com/google-ads/api/reference/rpc/v9/LocationInfo).

In the implementation we need to send two queries to Google Ads in order to get all the data of campaigns, campaign budgets, and campaign criterion:

1. Get campaigns and its budget in the first query.
2. Get campaign criterion with its corresponding campaign ID in the second query.
3. Combine two results in a single array by campaign ID.

This is because CampaignBudget is an Attributed Resource of Campaign, while CampaignCriterion is not an Attributed Resource of Campaign. More information can be seen [here](https://developers.google.com/google-ads/api/docs/query/structure#from). 

### Screenshots:

#### GET ads/campaigns

✍️ Note: it will be an old campaign created before this feature if the targeted_locations is an empty array.

<img width="1009" alt="Screenshot 2022-03-03 at 12 57 09" src="https://user-images.githubusercontent.com/914406/156499275-4904f66b-baa8-4304-bf08-ab722bafabfd.png">

#### GET ads/campaigns/:campaign-id - Has targeted locations

<img width="1010" alt="Screenshot 2022-03-03 at 12 58 32" src="https://user-images.githubusercontent.com/914406/156499372-296d5d61-947e-48d4-b868-a3ba43dc9ff3.png">

#### GET ads/campaigns/:campaign-id - No targeted locations

<img width="1008" alt="Screenshot 2022-03-03 at 12 57 57" src="https://user-images.githubusercontent.com/914406/156499413-57724d0e-f20b-43af-aa92-d2d3340b5e58.png">

### Detailed test instructions:

1. Create some paid campaigns in GLA's dashboard: `/wp-admin/admin.php?page=wc-admin&path=/google/dashboard`.
2. Go to Google Ads Campaigns Dashboard and add targeted locations in some of the campaigns you created.
    - <img width="2555" alt="Screenshot 2022-03-03 at 13 09 34" src="https://user-images.githubusercontent.com/914406/156500476-8d9e00d5-ff0f-429b-895c-3a4ffedae06d.png">
3. Using Postman, request to `GET ads/campaigns` and `GET ads/campaigns/:campaign-id` like the screenshots in the `Screenshots` section. Remember to add your `wp-cookie` and `x-wp-nonce` in the request headers.

### Additional details:

In this PR I haven't converted the `geo_target_constant` value returned from CampaignCriterion to `ISO 3166-1 alpha-2` country code format. I will update it in another PR once #1229 is done.

<!--
Optional.
Enter a summary of all changes in this Pull Request, which will be added to the changelog if accepted.
Each line should start with `(Fix|Add|Tweak|Update) - `, for example:
> Fix - I took care of something that wasn't working.
> Add - I added something new that's pretty cool.
> Tweak - I made a small change.
> Update - I made big changes to something that wasn't broken.

Leave the "Changelog entry" header in place completely empty, without any summary if no changelog entry is needed.
If you remove the "Changelog entry" header, the title of Pull Request will be used as the changelog entry.
-->
### Changelog entry

> Update - Include targeted locations in get campaigns requests

ianlin committed:
Include targeted locations in the response payload when get campaign(s)

- Query campaign criterion's location in a separate query
- Then combine two results using campaign ID

ianlin committed:
Add unit tests for ads multiple targeted locations

puntope commented on src/API/Google/AdsCampaign.php:
```diff
@@ -76,24 +78,30 @@ public function __construct( GoogleAdsClient $client, AdsCampaignBudget $budget
-			foreach ( $results->iterateAllElements() as $row ) {
-				$return[] = $this->convert_campaign( $row );
+			$campaign_criterion_results = ( new AdsCampaignCriterionQuery() )
+				->set_client( $this->client, $this->options->get_ads_id() )
+				->where( 'campaign.status', 'REMOVED', '!=' )
+				->get_results();
```
💅 


This chunk of code is repeated 2 times. The same happens in the `get_campaign` one 

What do you think of having a function like

```
function get_query_results( $query ) {
     return $query->set_client( $this->client, $this->options->get_ads_id() )
				->where( 'campaign.status', 'REMOVED', '!=' )
				->get_results();
}


$campaign_criterion_results = get_query_results( new AdsCampaignCriterionQuery() );
$campaign_results =  get_query_results( new AdsCampaignQuery() );
```

ianlin replied:
Good point, updated in f4a4342, thanks!

puntope commented on src/API/Google/AdsCampaign.php:
```diff
@@ -118,13 +126,26 @@ public function get_campaigns(): array {
+
+			foreach ( $campaign_results->iterateAllElements() as $row ) {
+				$converted_campaigns[] = $this->convert_campaign( $row );
+				break;
```
❓ Why it is necessary this break here?

ianlin replied:
The previous code before this PR was something like:

```js
foreach ( $campaign_results->iterateAllElements() as $row ) {
  return $this->convert_campaign( $row );
```

Which means we want to get the first element of `$campaign_results`, and since `$campaign_results` is an instance of `PagedListResponse` rather than an array, we couldn't use some of the array functions to retrieve the first element.

rawdreeg replied:
Maybe another way would be to move this piece of logic into a function that would take the  $campaign_results->iterateAllElements() array and return the **converted_campaigns**. You could have a second argument as a boolean flag (eg. **single** ) in order to return the first index of  **converted_campaigns** only using current() or reset().

```
foreach ( $campaign_results->iterateAllElements() as $row ) {
				$campaign                               = $this->convert_campaign( $row );
				$converted_campaigns[ $campaign['id'] ] = $campaign;
			}
```

puntope commented on src/API/Google/AdsCampaign.php:
```diff
@@ -388,6 +409,42 @@ protected function convert_campaign( GoogleAdsRow $row ): array {
+			// after https://github.com/woocommerce/google-listings-and-ads/issues/1229 is done.
+			$geo_target_constant = $location ? $location->getGeoTargetConstant() : null;
+
+			$campaign_key = array_search( $campaign_id, $campaign_ids, true );
+
+			if ( ! array_key_exists( 'targeted_locations', $campaigns[ $campaign_key ] ) ) {
+				$campaigns[ $campaign_key ]['targeted_locations'] = [];
+			}
+
+			if ( ! empty( $geo_target_constant ) ) {
+				$campaigns[ $campaign_key ]['targeted_locations'][] = $geo_target_constant;
+			}
```
❓ / 🚧  

I think we should check here if the `$campaign_key` was found because `array_search` will return false in the case is not found but the rest of the code in lines 436, 440 still runs so it will add all the targeted locations in `$campaigns[false] `

Since in PHP `$campaigns[false]`  evaluates to `$campaigns[0] ` it will create potential undesired behavioiur in this function.

ianlin replied:
Ah, that's indeed a potential bug, updated in 551c74b, thanks.

puntope reviewed (commented):
God work in this PR 👌 

Few comments regarding code to be discussed/addressed

ianlin committed:
Add a shared function for duplicate code in AdsCampaign query

ianlin committed:
Check if campaign id of criterion exists in campaigns

mikkamp commented:
I had a look at the code and I had another thought about how the two queries are combined that I wanted to mention.

With the query `AdsCampaignCriterionQuery` and the where condition of campaign.status != REMOVED we are querying all existing Campaign Criterion (which can potentially be a large set if they target many countries per campaign). For one that's going to make that `array_search` inefficient because we will repeat it many times.

But I was wondering if we should get the initial campaign query and store it based on campaign_id:
```php
$campaign = $this->convert_campaign( $row );
$converted_campaigns[ $campaign['id'] ] = $campaign;
```

We do something similar for reports and then before passing the results in the controller we use `array_values` to make it convert to a regular array (and match the schema).

This is not only going to remove the need for `array_search` but I think it will also be beneficial to move the `AdsCampaignCriterionQuery` within the `combine_campaigns_and_campaign_criterion_results` function. Since we know which set of campaigns we want data for, we can build the query with (when we get a single campaign there will only be one array key):

```php
->where( 'campaign.id', array_keys( $converted_campaigns ), 'IN' )
```

That would also make the query more flexible if we ever decide to display the campaigns in the UI with some pagination, or filtered to a subset of campaigns.

I'm also wondering whether we need some additional where conditions in the query. For example [`campaign_criterion.status`](https://developers.google.com/google-ads/api/fields/v9/campaign_criterion#campaign_criterion.status) has a possible status of `REMOVED` so if we edit the target countries and remove a country we previously had on the list. We'd still be getting it back as a result in the query, because the campaign hasn't been removed yet.
Also our extension isn't going to be using [`negative`](https://developers.google.com/google-ads/api/fields/v9/campaign_criterion#campaign_criterion.negative) criterion, but they could have created a campaign manually. Do you think it would be wise to only retrieve rows with `negative = false`, or should we keep those out some other way?

ianlin commented:
@mikkamp Thanks a lot for the suggestions!

> ->where( 'campaign.id', array_keys( $converted_campaigns ), 'IN' )

This is much more efficient in terms of querying campaign criterion, as I also noticed that for a campaign without setting targeted locations, there will still be three campaign criteria for that campaign returned from the campaign criterion query using current condition. I will update to what you suggested then, thanks again! 👍

> [campaign_criterion.status](https://developers.google.com/google-ads/api/fields/v9/campaign_criterion#campaign_criterion.status) has a possible status of REMOVED so if we edit the target countries and remove a country we previously had on the list.

Nice spotting! For the current design we're not allow the user to update the target countries once it has set, but in the future it could be so I'll still add this condition in the query tho, cheers!

> Also our extension isn't going to be using [negative](https://developers.google.com/google-ads/api/fields/v9/campaign_criterion#campaign_criterion.negative) criterion, but they could have created a campaign manually. Do you think it would be wise to only retrieve rows with negative = false, or should we keep those out some other way?

Another nice catch, I wasn't aware of this attribute but just like you said the user could have created a campaign via Google Ads console so I think it's a great idea to add the condition `negative = false` to retrieve only the "inclusive" criteria.

ianlin committed:
Enhance campaign criterion query performance

- Get only the campaign criteria by the campaign IDs we retrieved in the
	campaign query.
- No need to use array_search.
- Also add two conditions to campaign criterion query
  - campaign_criterion.negative = false (Whether to target (false) or exclude (true) the
		criterion.)
  - campaign_criterion.status != REMOVED

ianlin commented:
@mikkamp I've added a new commit cde88ee to enhance the campaign criterion query performance based on your suggestion. 😄

mikkamp commented on src/API/Site/Controllers/Ads/CampaignController.php:
```diff
@@ -262,41 +262,49 @@ function( &$value ) {
+				'items'       => [
+					'type' => 'string',
+				],
+			],
```
Once we change this to use the mapped values we can pass the same sanitation and validation functions as the country field. Also in the description I wouldn't mention it's coming from campaign criterion. The person consuming the API is only interested in a campaign and it's targeted locations, not how it's stored internally.

ianlin replied:
Noted, will add the sanitisation and validation functions in the next PR. Updated `target_locations`'s description in d5a6677.

mikkamp commented on src/API/Google/AdsCampaign.php:
```diff
@@ -429,4 +486,26 @@ protected function parse_campaign_id( string $name ): int {
+	 * @return PagedListResponse
+	 * @throws InvalidQuery When any condition array in conditions array is empty.
+	 */
+	protected function get_ads_query_results( AdsQuery $ads_query, array $conditions = [] ): PagedListResponse {
```
Just my personal opinion, but after the refactor it seems like this function is just adding more code and makes it harder to view what query is being built. We moved it down to just 3 times and the where clauses are very different, so the only part that repeats is `set_client( $this->client, $this->options->get_ads_id() )`. But happy to keep it in place if someone thinks otherwise.

If you do think it's good to keep this function then we don't need to reassign $ads_query. It returns an instance of itself to help with chaining, but if we call them individually we don't need to add that.

ianlin replied:
Agree, this helper method makes the where conditions harder to view. Updated in f7353d7.

mikkamp commented on src/API/Google/AdsCampaign.php:
```diff
@@ -388,6 +405,46 @@ protected function convert_campaign( GoogleAdsRow $row ): array {
+			}
+		}
+
+		return array_values( $campaigns );
```
I would personally have added array_values as the very last thing (not within this function). Since the index by campaign_id is added outside this function I think it makes more sense to remove it outside this function.

ianlin replied:
Agree, updated in 85bdb4e.

mikkamp commented on src/API/Google/AdsCampaign.php:
```diff
@@ -388,6 +405,46 @@ protected function convert_campaign( GoogleAdsRow $row ): array {
+			$geo_target_constant = $location ? $location->getGeoTargetConstant() : null;
+
+			if ( ! array_key_exists( 'targeted_locations', $campaigns[ $campaign_id ] ) ) {
+				$campaigns[ $campaign_id ]['targeted_locations'] = [];
```
The convert_campaign function already sets it to an empty array.

If I'd personally need to build in extra safety checks I'd rather skip the row if `$campaigns[ $campaign_id ]` is not set. It should never happen but it would seem more logical in case the query results are not what we expect.

ianlin replied:
Agree, updated in f5bacf6.

ianlin committed:
Update the schema description for targeted_locations

ianlin committed:
Remove the helper method get_ads_query_results()

Because:
- It actually adds more codes.
- It becomes harder to view the whole query.

ianlin committed:
Move `array_values` outside of data retrieving/transforming method

Since the index `campaign_id` was added outside the data
retrieving/transforming method, it makes more sense to remove it
outside of this method.

ianlin committed:
Check the key `campaign_id` is set or not rather than target_locations
array

- Since `target_locations` has already been set to an empty array in
	`convert_campaign` method.

ianlin commented:
@mikkamp Thanks for the review, I've updated the code and ready for another round of review.

mikkamp commented on src/API/Google/AdsCampaign.php:
```diff
@@ -118,13 +122,23 @@ public function get_campaigns(): array {
+			if ( ! empty( $converted_campaigns ) ) {
+				$combined_results        = $this->combine_campaigns_and_campaign_criterion_results( $converted_campaigns );
+				$combined_results_values = array_values( $combined_results );
+				return reset( $combined_results_values );
```
Reset will already return the first value within the array. So we don't need to pass it through array_values first.

ianlin replied:
Good catch, thanks for it! Since it's small so I updated it in 39a3a5e.

mikkamp commented on src/API/Google/AdsCampaign.php:
```diff
@@ -388,6 +403,44 @@ protected function convert_campaign( GoogleAdsRow $row ): array {
+	 * @return array
+	 */
+	protected function combine_campaigns_and_campaign_criterion_results( array $campaigns ): array {
+		$campaign_criterion_results = ( new AdsCampaignCriterionQuery() )->set_client( $this->client, $this->options->get_ads_id() )
+			->where( 'campaign.id', array_keys( $campaigns ), 'IN' )
+			// negative: Whether to target (false) or exclude (true) the criterion.
+			->where( 'campaign_criterion.negative', 'false', '=' )
+			->where( 'campaign_criterion.status', 'REMOVED', '!=' )
+			->get_results();
```
I added some location criterion directly in https://ads.google.com and I see them showing up. However I noticed that it also returns an additional `campaign_criterion` per campaign. We skip over them if the `$geo_target_constant` is empty, but would it be better to exclude them from the result set?

The best way I can see to do that is to add the following where condition to the query:
```
campaign_criterion.location.geo_target_constant IS NOT NULL
```

Query.php would need to be expanded to support the condition `IS NOT NULL`, but do you think that's worth it to add?
The query would be slightly larger, but we would only get the results we are interested in.

Should be possible to add with something like this (and allow it in validate_compare):
```php
			} elseif ( 'IS NOT NULL' === $compare ) {
				$value = '';
```

ianlin replied:
Nice, I think it's worth adding, it makes the query logic mostly in the SQL-like query. I'll refactor it in the "convert geo constants to county code" PR.

mikkamp reviewed (approved):
Thanks for all the refactoring so far. It looks good from my side and it's testing well.

I tested by adding some campaign locations in the dashboard (as well as some excluded locations). In the results I only saw the included locations showing up:

```json
{
	"id": 1234567890,
	"name": "Campaign 2022-01-02 12:00:00",
	"status": "enabled",
	"amount": 10,
	"country": "IE",
	"targeted_locations": [
		"geoTargetConstants/2372"
	]
}
```

I just left a minor comment about refactoring the query, but it's not an absolute must.

ianlin committed:
No need to get array values before using `reset`

Since `reset` already returns the first value of the array.
